src/ir/global-utils.h

```diff
@@ -62,9 +63,17 @@ inline bool canInitializeGlobal(const Expression* curr) {
     }
     return true;
   }
-  return Properties::isSingleConstantExpression(curr) ||
-         curr->is<GlobalGet>() || curr->is<RttCanon>() || curr->is<RttSub>() ||
-         curr->is<StructNew>();
+  if (Properties::isSingleConstantExpression(curr) || curr->is<GlobalGet>() ||
+      curr->is<RttCanon>() || curr->is<RttSub>() || curr->is<StructNew>() ||
+      curr->is<ArrayNew>() || curr->is<I31New>()) {
+    for (auto* child : ChildIterator(curr)) {
+      if (!canInitializeGlobal(child)) {
```

Comment:
Would it be worth making a one-off visitor here to avoid the recursion?

Reply:
Not sure what you mean by "one-off visitor"?

(If you mean to avoid recursion using an assumption about the depth, I think this can be pretty deep, e.g., struct.new with a field that contains another struct.new, repeated rtt.subs on each other, etc.)

Reply:
Sorry, I meant `Walker` when I wrote `Visitor`. I was thinking that we could use a `PostWalker` subclass to avoid the current unbounded recursion.

Reply:
Oh, I see. I agree that a Walker would be more efficient, but it would be more code and less clear IMO.

OTOH maybe that's a reason to try to make a more ergonomic walker for things like this... let me land this, and I'll think about the ergonomics. If I have a good idea there I can do a followup for this and other places in the code maybe.